### DISCUSSION TITLE
How do I implement email verification with this package?

### DISCUSSION THREAD

**Discussion by ervnjmsdnts** [fastapi-users/fastapi-users#800] [Q&A] [answered]
Generally after registration of the user; a confirmation email will be sent to them.

I tried implementing this by sending an email after the registration. But now it wants me to send the email of the user to /auth/request-verify-token route to get the token, then go to /auth/verify to verify the token and set the user to be verified.

How will I implement this when I can't send a post request internally???

BTW: Update your docs specifically in the functionality of overriding the User Manager methods

**Comment by ervnjmsdnts** [comment, marked as answer]
After some tinkering with the API the solution that I manage to create is to make a new route that would just update the is_verified field when the email is verified

**Reply by junoriosity** [reply]
Hi @ervnjmsdnts could you share your solution here, such that I can try it as well?

**Reply by ervnjmsdnts** [reply]
Oh I forgot to post the solution sorry @junoriosity

**Reply by ervnjmsdnts** [reply]
Here is the solution: 
```
class UserManager(BaseUserManager[UserCreate, UserDB]):
    user_db_model = UserDB
    verification_token_secret = VERIFY_TOKEN_SECRET

    async def validate_password(self, password: str, user: Union[UserCreate, UserDB]) -> None:
        if password != user.confirmationPassword:
            raise InvalidPasswordException(reason="Passwords don't match")

    async def on_after_register(self, user: UserDB, request: Optional[Request] = None):
        payload = {"email": user.email}
        token = jwt.encode(payload, self.verification_token_secret)
        await sendEmail("Email Confirmation", user.email, token)
```
Here is the User Manager. I override the on_after_register method that would send a token housing the user's email as the payload using the sendEmail function.

```
async def verifyEmail(token: str):
    try:
        payload = jwt.decode(token, VERIFY_TOKEN_SECRET, algorithms="HS256")
        await user_collection.update_one({"email": payload["email"]}, {"$set": {"is_verified": True}})

        return True

    except:
        return False
```
Here is my function that would set the user's is_verified field to True.
```
@auth_router.get("/confirmation/{id}")
async def confirmation(id: str):
    result = await verifyEmail(id)

    if result:
        return {"message": "Email is confirmed"}

    return {"message": "Email not confirmed"}
```
Then lastly here is the confirmation route.

**Comment by hgalytoby** [comment]
source code  `/auth/request-verify-token` use 29 lines  `user_manager.request_verify(user, request)`

https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/router/verify.py#L11

Copy lines 288 to 297

https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/manager.py#L269

```py
from fastapi_users.jwt import generate_jwt


class UserManager(UUIDIDMixin, BaseUserManager[UserModel, uuid.UUID]):
    reset_password_token_secret = SECRET
    verification_token_secret = SECRET

    async def on_after_register(self, user: UserModel, request: Optional[Request] = None):
        token_data = {
            "user_id": str(user.id),
            "email": user.email,
            "aud": self.verification_token_audience,
        }
        token = generate_jwt(
            token_data,
            self.verification_token_secret,
            self.verification_token_lifetime_seconds,
        )
        sendEmail("Email Confirmation", user.email, token)
```
